src/sync/barrier.rs

```diff
@@ -56,57 +67,77 @@ impl Barrier {
-            state.leader = Some(me);
-        }
+        let my_epoch = state.epoch;
+
+        trace!(waiters=?state.waiters, epoch=my_epoch, "waiting on barrier {:p}", self);
 
         // Update the barrier's clock with the clock of this thread
         ExecutionState::with(|s| {
             let clock = s.increment_clock();
             state.clock.update(clock);
         });
 
-        if state.waiters.len() + 1 < state.bound {
-            // Block current thread
-            assert!(state.waiters.insert(me)); // current thread shouldn't already be in the set
+        // The current thread shouldn't already be in `waiters`.
+        assert!(state.waiters.insert(me));
+
+        if state.waiters.len() < state.bound {
+            trace!(waiters=?state.waiters, epoch=my_epoch, "blocked on barrier {:?}", self);
             ExecutionState::with(|s| s.current_mut().block(false));
-            trace!(leader=?state.leader, waiters=?state.waiters, "blocked on barrier {:?}", self);
         } else {
-            trace!(leader=?state.leader, waiters=?state.waiters, "releasing waiters on barrier {:?}", self);
-            // Unblock each waiter, updating its clock with the barrier's clock
+            trace!(waiters=?state.waiters, epoch=my_epoch, "releasing waiters on barrier {:?}", self);
+
+            debug_assert!(state.waiters.len() == state.bound || state.bound == 0);
+
+            // Make the leader token available for this epoch. The first task to wake up will
+            // take it and become the leader. The token shouldn't already be available.
+            assert!(state.leader_tokens.insert(my_epoch));
+
+            // Drain the set of waiters and increment the barrier's epoch, so any other task that
+            // calls `wait` from now on becomes part of a separate group with its own leader.
+            let waiters: Vec<TaskId> = state.waiters.drain().collect();
+            state.epoch += 1;
+
+            trace!(
+                waiters=?state.waiters,
+                epoch=state.epoch,
+                "releasing waiters on barrier {:?}",
+                self,
+            );
+
             let clock = state.clock.clone();
             ExecutionState::with(|s| {
-                for tid in state.waiters.drain() {
-                    debug_assert_ne!(tid, me);
+                // `waiters` includes the current task.
+                for tid in waiters {
                     let t = s.get_mut(tid);
-                    debug_assert!(t.blocked());
                     t.clock.increment(tid);
                     t.clock.update(&clock);
-                    t.unblock();
+
+                    // A thread blocked on the barrier will be blocked iff it's not the current
+                    // one, since this is the last thread arriving at the barrier for this epoch.
+                    assert_eq!(t.blocked(), tid != me);
+                    if t.blocked() {
+                        t.unblock();
+                    }
```

Comment:
Don't think this needs the guard (`unblock` doesn't check).
```suggestion
                    t.unblock();
```